### ISSUE TITLE
UnreadFields detector false positives on protected fields when using onlyAnalyze

### ISSUE THREAD

**Issue by Vogel612** [spotbugs/spotbugs#2700]
The MCVE for this is slightly more complex, so I made it available in a separate repository: https://github.com/Vogel612/mcve-spotbugs-unread-fields

The gist of the issue is that when an analysis is restricted to a Child class using `onlyAnalyze`, the `UnreadFields` detector seems to be unaware of field initializers for inherited fields.
This results in a report without `onlyAnalyze` showing no results, while the report that is restricted to fewer classes generates false positive results.
This is true both for trivial field initializers as well as invocations of factory methods:

```java
public class Parent {
    protected final Factory factory = Factory.instance();
    protected final Object comparison = new Object();
}
---
public class Inheritor extends Parent {
    public Inheritor() {}
    public void usage() {
        this.factory.doSomething();
        System.out.println(this.comparison.hashCode());
    }
}
```

Both cases generate an incorrect `NP_UNWRITTEN_PUBLIC_OR_PROTECTED_FIELD` warning with the `onlyAnalyze` option while an unrestricted analysis does not generate this warning.

I verified this behavior with Spotbugs 4.8.1 as well as 4.7.3.

**Comment by gtoison** [comment]
I looked into it and (unless I misunderstood) the other classes are simply ignored when using `onlyAnalyze`.
Since SpotBugs has no knowledge of the other classes (including parent classes), false positives (and false negatives) will occur.
The detectors also do not know that SpotBugs is running with `onlyAnalyze` so they cannot easily disable themselves knowing that some classes were ignored.

So I'm not sure how solve this, IMO it is a short-coming of `onlyAnalyze`
